src/http/server.cr

```diff
@@ -95,72 +100,187 @@ class HTTP::Server
   end
 
-  # Returns the TCP port the server is connected to.
+  # Creates a `TCPServer` and adds it as an interface.
+  #
+  # If *port* is `0`, a random, free port will be chosen.
+  #
+  # You may set *reuse_port* to `true` to enable the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind(host : String, port : Int32, reuse_port : Bool = false) : TCPServer
+    bind TCPServer.new(host, port, reuse_port: reuse_port)
+  end
+
+  # Creates a `TCPServer` listenting on `127.0.0.1` and adds it as an interface.
+  #
+  # If *port* is `0`, a random, free port will be chosen.
+  #
+  # You may set *reuse_port* to true to enable the `SO_REUSEPORT` socket option,
+  # which allows multiple processes to bind to the same port.
+  def bind(port : Int32, reuse_port : Bool = false) : TCPServer
+    bind "127.0.0.1", port, reuse_port
+  end
+
+  # Creates a `TCPServer` listening on an unused port and adds it as an interface.
+  #
+  # Returns the determined port number.
+  #
+  # ```
+  # server = HTTP::Server.new { }
+  # server.bind_unused_port # => 12345
+  # ```
+  def bind_unused_port(host : String = "127.0.0.1", reuse_port : Bool = false) : Int32
+    tcp_server = bind(host, 0, reuse_port)
+    tcp_server.local_address.port
+  end
+
+  # Creates a `Socket::Server` listenting to *address* and adds it as an interface.
+  #
+  # The *address* can either be `host:port` or a Unix socket prefixed by `unix:`.
+  # This obviously works only on systems supporting Unix sockets.
   #
-  # For example you may let the system choose a port, then report it:
   # ```
-  # server = HTTP::Server.new(0) { }
-  # server.bind
-  # server.port # => 12345
+  # server = HTTP::Server.new { }
+  # server.bind "127.0.0.1:8080"
+  # server.bind "unix:/tmp/my-socket"
   # ```
-  def port
-    if server = @server
-      server.local_address.port.to_i
+  def bind(address : String, reuse_port : Bool = false) : Socket::Server
+    if address.starts_with?("unix:")
+      {% if flag?(:unix) %}
+      bind(Socket::UNIXAddress.new(address[5..-1])).as(Socket::Server)
+      {% else %}
+      raise NotImplementedError.new "system does not support Unix sockets"
+      {% end %}
     else
-      @port
+      host, _, port = address.partition(':')
+      bind(host, port.to_i, reuse_port).as(Socket::Server)
     end
   end
 
-  # Creates the underlying `TCPServer` if the doesn't already exist.
+  {% if flag?(:unix) %}
+  # Creates a `UNIXServer` bound to *address* and adds it as an interface.
   #
-  # You may set *reuse_port* to true to enable the `SO_REUSEPORT` socket option,
-  # which allows multiple processes to bind to the same port.
-  def bind(reuse_port = false)
-    @server ||= TCPServer.new(@host, @port, reuse_port: reuse_port)
+  # ```
+  # server = HTTP::Server.new { }
+  # server.bind Socket::UNIXAddress.new("/tmp/my-socket")
+  # ```
+  def bind(address : Socket::UNIXAddress) : UNIXServer
```

Comment:
Are the XAddress overloads *really* useful?

Reply:
I guess we don't necessarily need them, it's just that the Socket API is not really nice in that regard, you currently can't directly create a UNIXSocket from a UNIXAddress. But this needs to be fixed there (see #5778) then we can probably do without the overloads. Or have only a general `bind(address : Socket::Address)` which delegates creating the appropriate server to `Socket::Server.new(address : Socket::Address)`. That would probably be the best solution.